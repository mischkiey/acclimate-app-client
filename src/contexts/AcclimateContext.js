import React from 'react';

const AcclimateContext = React.createContext({});

export default AcclimateContext;