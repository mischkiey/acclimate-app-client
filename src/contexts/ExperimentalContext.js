import React, { Component } from 'react';
import TokenService from '../services/token-services';

export const ExperimentalContext = React.createContext({});

export class ExperimentalProvider extends Component {
    state = {
        log: false,
        tasks: [],
        shoppingItems: [],
    };

    handleUserLog = () => {
        const log = TokenService.hasAuthToken();
        this.setState({log});
    };

    handleUserLogOut = () => {
        TokenService.clearAuthToken();
        this.handleUserLog();
    };

    setTasks = (tasks) => {
        this.setState({tasks})
    };

    setShoppingItems = (shoppingItems) => {
        this.setState({shoppingItems})
    };

    render() {
        const value = {
            log: this.state.log,
            tasks: this.state.tasks,
            shoppingItems: this.state.shoppingItems,

            handleUserLog: this.handleUserLog,
            handleUserLogOut: this.handleUserLogOut,
            setTasks: this.setTasks,
            setShoppingItems: this.setShoppingItems,
        };

        return (
            <ExperimentalContext.Provider value={value}>
                {this.props.children}
            </ExperimentalContext.Provider>
        );
    };
};

