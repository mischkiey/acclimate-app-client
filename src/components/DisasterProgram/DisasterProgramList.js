import React, {Component} from 'react';
import APIService from '../../services/api-services';
import TokenService from '../../services/token-services';
import { ExperimentalContext } from '../../contexts/ExperimentalContext';

import DisasterProgram from '../../components/DisasterProgram/DisasterProgram';

class DisasterProgramList extends Component {
    static contextType = ExperimentalContext;

    state = {
        programs: [],
        error: null,
    };

    handleAddToTaskList = (disaster_plan_step_itemable_shorthand) => {
        this.setState({error: null})

        const newUserTask = {
            user_task: disaster_plan_step_itemable_shorthand,
        };
        const token = TokenService.getAuthToken();

        return APIService.post('/disaster/user/task', newUserTask, token)
            .then(response => {
                this.context.tasks.push(response);
                const newTasks = this.context.tasks;
                this.context.setTasks(newTasks);
            })
            .catch(error => {
                this.setState({...error});
            });
    };

    handleAddToShoppingList = (disaster_plan_step_itemable_shorthand) => {
        this.setState({error: null})

        const newUserShoppingItem = {
            user_shopping_item: disaster_plan_step_itemable_shorthand,
        };
        const token = TokenService.getAuthToken();

        return APIService.post('/disaster/user/shopping', newUserShoppingItem, token)
            .then(response => {
                this.context.shoppingItems.push(response);
                const newShoppingItems = this.context.shoppingItems;
                this.context.setShoppingItems(newShoppingItems);
            })
            .catch(error => {
                this.setState({...error});
            });
    };

    handleDeleteDisasterProgram = (disaster_program_id) => {
        this.setState({error: null})

        const token = TokenService.getAuthToken();

        return APIService.del(`/disaster/user/program/${disaster_program_id}`, token)
            .then(() => {
                return APIService.get('/disaster/user/program', token)
                    .then(programs => {
                        this.setState({programs})
                    })
                    .catch(error => {
                        this.setState({error})
                    })
            })
            .catch(error => {
                this.setState({...error})
            });
    };

    componentDidMount() {
        this.setState({error: null})

        const token = TokenService.getAuthToken();

        return APIService.get('/disaster/user/program', token)
            .then(programs => {
                this.setState({programs})
            })
            .catch(({error}) =>{
                this.setState({error})
            });
    };

    render () {
        const programs = this.state.programs.map(program =>
            <DisasterProgram
                key={program.disaster_program_id}
                handleAddToTaskList={this.handleAddToTaskList}
                handleAddToShoppingList={this.handleAddToShoppingList}
                handleDeleteDisasterProgram={this.handleDeleteDisasterProgram}
                {...program}
            />
        )

        return (
            <section className='dashboard-component-group item'>
                {programs}
            </section>
        );
    };
};

export default DisasterProgramList;