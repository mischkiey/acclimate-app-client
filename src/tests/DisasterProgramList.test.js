import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { AcclimateProvider } from '../contexts/AcclimateContext';
import DisasterProgramList from '../components/DisasterProgram/DisasterProgramList';

describe(`DisasterProgramList Component`, () => {
    describe(`Smoke test`, () => {
        it(`Renders without crashing`, () => {
            const div = document.createElement('div');
            ReactDOM.render(
                <BrowserRouter>    
                    <AcclimateProvider>
                        <DisasterProgramList />
                    </AcclimateProvider>
                </BrowserRouter>
                , div);
            ReactDOM.unmountComponentAtNode(div);
        });
    });
    
    describe(`Snapshot test`, () => {
        it(`Renders the UI as expected`, () => {
            const tree = renderer
                .create(
                    <BrowserRouter>    
                        <AcclimateProvider>
                            <DisasterProgramList />
                        </AcclimateProvider>
                    </BrowserRouter>
                )
                .toJSON()
            expect(tree).toMatchSnapshot()
        });
    });
})