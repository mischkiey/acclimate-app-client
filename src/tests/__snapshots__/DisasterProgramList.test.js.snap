// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisasterProgramList Component Snapshot test Renders the UI as expected 1`] = `
<section
  className="dashboard-components-group item"
/>
`;
